frame/im-online/src/lib.rs

```diff
@@ -315,13 +315,32 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		#[weight = MINIMUM_WEIGHT]
+		/// `keys_len` must be the length for the vec stored in storage `Keys`.
+		///
+		/// # <weight>
+		/// - Complexity: `O(K + E)` where K is length of `Keys` and E is length of
+		///   `Heartbeat.network_state.external_address`
+		///
+		///   - `O(K)`: decoding of length `K`
+		///   - `O(E)`: decoding/encoding of length `E`
+		/// - DbReads: pallet_session `Validators`, pallet_session `CurrentIndex`, `Keys`,
+		///   `ReceivedHeartbeats`
+		/// - DbWrites: `ReceivedHeartbeats`
+		/// # </weight>
+		// NOTE: the weight include cost of validate_unsigned as it is part of the cost to import
+		// block with such an extrinsic.
+		#[weight = 310_000_000
+			+ 750_000 * *_keys_len as Weight
+			+ 1_200_000 * heartbeat.network_state.external_addresses.len() as Weight
+			+ T::DbWeight::get().reads_writes(4, 1)]
 		fn heartbeat(
 			origin,
 			heartbeat: Heartbeat<T::BlockNumber>,
 			// since signature verification is done in `validate_unsigned`
 			// we can skip doing it here again.
-			_signature: <T::AuthorityId as RuntimeAppPublic>::Signature
+			_signature: <T::AuthorityId as RuntimeAppPublic>::Signature,
+			// checked in `validate_unsigned`
+			_keys_len: u32,
```

Comment:
Here keys_len isn't part of the signed heartbeat but I think its fine, though one can double-check.

Reply:
Since imonline transactions have to be propagated to other authorities (since the ones that send it didn't have a chance to produce a block), there is a potential attack, where other authorities change the `key_len` inside the payload making the transaction invalid.
This is not super bad, since they can also change say `session_index` or simply censor the transactions, but with `key_len` we can't be sure what happened - we don't know if the authority is misbehaving or it was maliciously modified. The possible cases are:
1. The hearbeat is incorrect, but has correct signature (i.e. the authority is misbehaving).
2. The heartbeat is incorrect and has incorrect signature (i.e. was modified).
3. The heartbeat was not seen (i.e. either not sent or censored).
4. `Key_len` invalid (i.e. either modified or authority misbehaving).

I'd be for putting that into `Heartbeat` payload for simplicity, unless there are any reasons why we should not do that.

Reply:
Re discussions in benchmarking channel, we should probably use saturating ops here, since `key_len` is coming from the user.

Even though the transaction should be rejected, I think the overflow may have some unforseen consequences (for instance it will panic in debug mode), so better have saturating ops for everything here.

Reply:
ok for the saturation, params should have been checked in validate_unsigned but better be safe anyway.

I updated to put key_len part of heartbeat. :+1: